### PR TITLE
feat: Improve bump_version() behavior in the absence of changes

### LINKED ISSUE #307: Add noop argument (or similar) to bump_version()
```r
bump_version <- function(which = "dev", noop = c("bump", "noop", "fail")) {
  ...
}
```

### LINKED ISSUE #316: where does range come from
https://github.com/cynkra/fledge/blob/520b9b2e36da7bc58136378407f1f4454bfbeadf/R/update-news.R#L25

### PR DESCRIPTION
**Pull request by maelle**
Fix #316 
Fix #307

### CODE REVIEW

File: R/api-update-news.R [hunk 1]
```diff
@@ -6,16 +6,23 @@
...
-    messages <- get_top_level_commits(since = get_last_tag()$commit)$message
+    commit_range <- commit_range %||% default_commit_range()
+    messages <- commit_range$message
   }
 
   with_repo(update_news_impl(messages))
   invisible(NULL)
 }
+
+default_commit_range <- function() {
```
Comment (maelle) [inline R/api-update-news.R:26]: my change means the last tag is not found, I am investigating

File: tests/testthat/_snaps/bump_version.md [hunk 2]
```diff
@@ -23,6 +23,44 @@
...
+    Output
+      NULL
+
+---
+
+    Code
+      bump_version(no_change_behavior = "bump")
+    Message
+      > Scraping 1 commit messages.
+      v Found 1 NEWS-worthy entries.
```
Comment (maelle) [inline tests/testthat/_snaps/bump_version.md:45]: this is not ok, nothing should be worthy

File: tests/testthat/test-bump_version.R [hunk 3]
```diff
@@ -13,6 +13,15 @@ test_that("bump_version() works -- dev", {
   expect_equal(as.character(desc::desc_get_version()), "0.0.0.9001")
   expect_equal(get_last_tag()$name, "v0.0.0.9001")
   expect_snapshot_file("NEWS.md", compare = compare_file_text)
+
+  ## no changes
+  expect_snapshot_error(bump_version(no_change_behavior = "fail"), variant = snapshot_variant("testthat"))
+  expect_snapshot(bump_version(no_change_behavior = "noop"), variant = snapshot_variant("testthat"))
+  expect_equal(as.character(desc::desc_get_version()), "0.0.0.9001")
+  expect_equal(get_last_tag()$name, "v0.0.0.9001")
```
Comment (maelle) [inline tests/testthat/test-bump_version.R:21]: here the tag is found.
it is however not found in `update_news()`.
I'm looking into `local_demo_project()` a bit more.

File: R/api-bump-version.R [hunk 4]
```diff
@@ -16,8 +20,11 @@
 #' Fear not, run [unbump_version()], merge that PR, run `bump_version()`.
 #'
 #' @example man/examples/bump-version.R
-bump_version <- function(which = "dev") {
+bump_version <- function(which = "dev", no_change_behavior = c("bump", "noop", "fail")) {
   check_which(which)
+
+  no_change_behavior <- match.arg(no_change_behavior, choices = c("bump", "noop", "fail"))
```
Comment (krlmlr) [inline R/api-bump-version.R:26]: `rlang::arg_match()` gives rlang-style error messages.

File: R/bump-version.R [hunk 5]
```diff
@@ -1,9 +1,24 @@
 #' @rdname bump_version
 #' @usage NULL
-bump_version_impl <- function(which) {
+bump_version_impl <- function(which, no_change_behavior) {
   #' @description
   #' 1. Verify that the current branch is the main branch.
   check_main_branch()
+  #' 1. Check there were changes since the last version.no_change()
```
Comment (krlmlr) [inline R/bump-version.R:7]:
```suggestion
  #' 1. Check if there were changes since the last version.
```

Comment (krlmlr) [inline R/bump-version.R:7]:
```suggestion
  #' 1. Check if there were changes since the last version.
```

File: R/api-bump-version.R (hunk 4 above)
Comment (krlmlr) [inline R/api-bump-version.R:26]: `rlang::arg_match()` gives rlang-style error messages, can we pick up the options from the formals?


```suggestion
  no_change_behavior <- arg_match(no_change_behavior)
```

Comment (krlmlr) [inline R/api-bump-version.R:26]: `rlang::arg_match()` gives rlang-style error messages, can we pick up the options from the formals?


```suggestion
  no_change_behavior <- arg_match(no_change_behavior)
```

File: R/bump-version.R [hunk 6]
```diff
@@ -78,3 +93,7 @@ get_main_branch_config <- function() {
   global <- init[init$level == "global"]
   return(global$value)
 }
+
+no_change <- function() {
+  nrow(default_commit_range()) == 1
```
Comment (krlmlr) [inline R/bump-version.R:98]: Why 1? Can you please add a comment? Should we do `<= 1` for safety?

Comment (krlmlr) [inline R/bump-version.R:98]: Why 1? Can you please add a comment?

File: R/update-news.R [hunk 7]
```diff
@@ -22,7 +22,7 @@ collect_news <- function(messages) {
     unlist()
 
   if (length(message_items) == 0) {
-    if (length(range) <= 1) {
+    if (length(messages) <= 1) {
```
Comment (krlmlr) [inline R/update-news.R:25]: The logic seems repeated here. Should `no_change()` get a `messages` argument?

Comment (krlmlr) [inline R/update-news.R:25]: The logic seems repeated here. Should `no_change()` get a `messages` argument?

File: tests/testthat/test-bump_version.R [hunk 8]
```diff
@@ -13,6 +13,18 @@ test_that("bump_version() works -- dev", {
...
+  expect_snapshot(bump_version(no_change_behavior = "bump"), variant = snapshot_variant("testthat"))
+  expect_equal(as.character(desc::desc_get_version()), "0.0.0.9002")
+  expect_equal(get_last_tag()$name, "v0.0.0.9002")
+  file.copy("NEWS.md", "NEWS2.md")
+
+  expect_snapshot_file("NEWS2.md", compare = compare_file_text)
```
Comment (krlmlr) [inline tests/testthat/test-bump_version.R:27]: Can we ask testthat to do the file copying?

```suggestion
  expect_snapshot_file("NEWS.md", "NEWS2.md", compare = compare_file_text)
```

Comment (krlmlr) [inline tests/testthat/test-bump_version.R:27]: Can we ask testthat to do the file copying?

```suggestion
  expect_snapshot_file("NEWS.md", "NEWS2.md", compare = compare_file_text)
```

File: tests/testthat/_snaps/bump_version.md [hunk 9]
```diff
@@ -23,6 +23,44 @@
...
+    x No change since last version.
+    i Use `no_change_behavior = "bump"` to force a version bump, or
+              `no_change_behavior = "noop"` to do nothing.
+
+---
+
+    Code
+      bump_version(no_change_behavior = "noop")
+    Output
+      NULL
```
Comment (krlmlr) [inline tests/testthat/_snaps/bump_version.md:37]: Do we want an informative message here?

```suggestion
        i No change since last version.
```

Comment (krlmlr) [inline tests/testthat/_snaps/bump_version.md:37]: Do we want an informative message here?

```suggestion
        i No change since last version.
```

**Review by krlmlr — COMMENTED** [review]
Thanks, I like the tests a lot. Let's do another pass.

File: R/update-news.R (hunk 7 above)
Comment (maelle) [reply R/update-news.R:25]: I've changed the variable names to clarify that we are looking for `newsworthy_items` among `messages`.
Even if `no_change()` was run before getting to this point, if the argument `no_change_behavior` could mean we arrive at this point.
But maybe in the case there's less than or 1 message the function should exit earlier, is this what you mean?

File: R/update-news.R [hunk 10]
```diff
@@ -14,26 +14,26 @@ collect_news <- function(messages) {
...
+      newsworthy_items <- "- Same as previous version."
     } else {
-      message_items <- "- Internal changes only."
+      newsworthy_items <- "- Internal changes only."
     }
   }
 
   if (fledge_chatty()) {
-    cli_alert_success("Found {.field {length(message_items)}} NEWS-worthy entries.")
+    cli_alert_success("Found {.field {length(newsworthy_items)}} NEWS-worthy entries.")
```
Comment (maelle) [inline R/update-news.R:33]: see #325 as this can be confusing (to be fixed in a later PR)

File: tests/testthat/_snaps/bump_version.md [hunk 11]
```diff
@@ -55,3 +95,7 @@
     x Must be on the main branch (main) for running fledge functions.
     i Currently on branch bla.
 
+# bump_version() errors well for wrong arguments
```
Comment (maelle) [inline tests/testthat/_snaps/bump_version.md:98]: This reminds me of #153 (the error for the which argument looks different at the moment)

File: R/update-news.R (hunk 7 above)
Comment (krlmlr) [reply R/update-news.R:25]: Should we have `no_change()` handle the `<= 1` comparison, so that this logic is centralized, if it makes sense?

**Review by krlmlr — APPROVED** [review]
Thanks, looks great!

**Pull request merged**